src/rapids_singlecell/preprocessing/_kernels/_bbknn.py

```diff
@@ -0,0 +1,96 @@
+from __future__ import annotations
+
+import cupy as cp
+
+_find_top_k_kernel = r"""
+extern "C" __global__
+void find_top_k_per_row(
+    const float* __restrict__ data,
+    const int* __restrict__ indptr,
+    const int n_rows,
+    const int trim,
+    float* __restrict__ vals) {
+
+    extern __shared__ float shared_memory[];
```

Comment:
hmm, cuda guarantees that to be 32 bit right? https://docs.nvidia.com/cuda/npp/nppdefs.html#c.Npp32f

regarding performance: does cuda do something to avoid race conditions for shared memory? because you don’t need that.